llvm/lib/Target/RISCV/RISCVFrameLowering.cpp

```diff
@@ -1502,6 +1659,9 @@ void RISCVFrameLowering::determineCalleeSaves(MachineFunction &MF,
   auto *RVFI = MF.getInfo<RISCVMachineFunctionInfo>();
   if (RVFI->isPushable(MF) && SavedRegs.test(RISCV::X26))
     SavedRegs.set(RISCV::X27);
+
+  // SiFive Preemptible Interrupt Handlers need additional frame entries
+  createSiFivePreemptibleInterruptFrameEntries(MF, *RVFI);
```

Comment:
Should this be called from `assignCalleeSavedSpillSlots` rather than `determineCalleeSaves`. `assignCalleeSavedSpillSlots` is where other stack slot objects are created.

Reply:
This has to be called here. `assignCalleeSavedSpillSlots` in PrologEpilogInserter.cpp  will return before calling the target-specific `assignCalleeSavedSpillSlots` if `SavedRegs` (the bitmap of registers to save) is empty, which it could be. Because we don't model mepc/mcause state as a register, we don't have a bit in the bitmap to use to say we're saving it.

Reply:
Do we need to consider these frame indices in the `MinCSFrameIndex` and `MaxCSFrameIndex` returned by `assignCalleeSavedSpillSlots`?

Reply:
I'm not entirely sure. Those two are used to work out which are conventional callee-saves when sorting the frame objects. If I include the two additional frame objects, it entirely reorders the callee-saves on the stack, which might be ok, but it's hard to tell. The two variables are also used in `computeFreeStackSlots`, which is what enabled #125959 - so it's probably best that we mark these as live as the other callee-saved frame objects.